model-optimizer/extensions/ops/gatherelements.py

```diff
@@ -0,0 +1,74 @@
+"""
+ Copyright (C) 2017-2020 Intel Corporation
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+"""
+
+import numpy as np
+
+from mo.graph.graph import Node, Graph
+from mo.ops.op import Op, PermuteAttrs
+
+
+class GatherElements(Op):
+    op = 'GatherElements'
+
+    def __init__(self, graph: Graph, attrs: dict):
+        super().__init__(graph, {
+            'op': self.op,
+            'type': self.op,
+            'version': 'opset6',
+            'infer': self.infer,
+            'in_ports_count': 2,
+            'out_ports_count': 1,
+            'axis': 0,
+        }, attrs)
+
+    def backend_attrs(self):
+        return ['axis']
+
+    @staticmethod
+    def infer(node: Node):
+        data_shape = node.in_port(0).data.get_shape()
+        indices_shape = node.in_port(1).data.get_shape()
+        axis = node.axis
+        data_rank = len(data_shape)
+
+        assert data_rank == len(indices_shape), 'data and indices inputs for node {} must be of the ' \
+                                                'same rank. Instead got {} and {}'.\
+                                                format(node.name, data_rank, len(indices_shape))
+        assert -data_rank <= axis < data_rank, 'axis for node {0} must be within interval ' \
+                                               '[-{1}},  {1} - 1]. Instead got: axis={2}'.\
+                                               format(node.name, data_rank, axis)
+        if axis < 0:
+            axis += data_rank
+
+        for idx, (data_sz, ind_sz) in enumerate(zip(data_shape, indices_shape)):
+            if idx != axis and data_sz != ind_sz:
+                raise ValueError('Sizes along axis {} for node {} do not match. data and indices must have ' 
+                                 'equal size along all axes except for axis {}'.format(idx, node.name, axis))
+
+        data = node.in_port(0).data.get_value()
+        indices = node.in_port(1).data.get_value()
+
+        if data is not None and indices is not None:
+            out_value = np.empty(indices_shape, dtype=data.dtype)
+            for idx in np.ndindex(*indices_shape):
+                data_idx = list(idx)
+                data_idx[node.axis] = indices[idx]
+                out_value[idx] = data[tuple(data_idx)]
+            node.out_port(0).data.set_value(out_value)
+        else:
+            node.out_port(0).data.set_shape(indices_shape)
+
+        PermuteAttrs.create_permute_attrs(node, attrs=[('axis', 'input:0')])
```

Comment:
@lazarevevgeny, @pavel-esir, why can `indices` tensor not be of a rank greater 3?

Reply:
Sorry, but I don't get your question. This line is a standard approach to permute value for the attribute specifying axis. There are no constraints to the rank of `indices`. It could be of any dimension. But if it will be 4D or 5D then the value for attribute "axis" will be updated accordingly.

Reply:
Sorry, I seem to throw a wrong question. Let me clarify. Is `indices` tensor transposed in case its rank greater 3? If yes, why is the calculation still valid then?

Reply:
We expect that it has been already transposed by some other transformation. For example, in original model this input was produced by the Convolution node which will produce output in NHWC layout. During the MO conversion the output from Convolution node will be in NCHW layout, so the `axis` must be updated. Of course, there could be a more complex situation when the `indices` input is not transposed so the `axis` update will be wrong. But such a case should be handled separately. But our default implementation is to mark such attributes to be transposed.

Reply:
I got it. Thanks. It is from ONNX specification so I see no problem then.